libndt.hpp

```diff
@@ -836,6 +836,63 @@ class Client {
 // the interface and implementation using LIBNDT_NO_INLINE_IMPL.
 #ifndef LIBNDT_NO_INLINE_IMPL
 
+#ifdef __linux__
+#include <linux/tcp.h>
+#define NDT7_ENUM_TCP_INFO \
+XX(tcpi_state, TcpiState) \
+XX(tcpi_ca_state, TcpiCaState) \
+XX(tcpi_retransmits, TcpiRetransmits) \
+XX(tcpi_probes, TcpiProbes) \
+XX(tcpi_backoff, TcpiBackoff) \
+XX(tcpi_options, TcpiOptions) \
+XX(tcpi_snd_wscale, TcpiSndWscale) \
+XX(tcpi_rcv_wscale, TcpiRcvWscale) \
+XX(tcpi_delivery_rate_app_limited, TcpiDeliveryRateAppLimited) \
+XX(tcpi_rto, TcpiRto) \
+XX(tcpi_ato, TcpiAto) \
+XX(tcpi_snd_mss, TcpiSndMss) \
+XX(tcpi_rcv_mss, TcpiRcvMss) \
+XX(tcpi_unacked, TcpiUnacked) \
+XX(tcpi_sacked, TcpiSacked) \
+XX(tcpi_lost, TcpiLost) \
+XX(tcpi_retrans, TcpiRetrans) \
+XX(tcpi_fackets, TcpiFackets) \
+XX(tcpi_last_data_sent, TcpiLastDataSent) \
+XX(tcpi_last_ack_sent, TcpiLastAckSent) \
+XX(tcpi_last_data_recv, TcpiLastDataRecv) \
+XX(tcpi_last_ack_recv, TcpiLastAckRecv) \
+XX(tcpi_pmtu, TcpiPmtu) \
+XX(tcpi_rcv_ssthresh, TcpiRcvSsthresh) \
+XX(tcpi_rtt, TcpiRtt) \
+XX(tcpi_rttvar, TcpiRttvar) \
+XX(tcpi_snd_ssthresh, TcpiSndSsthresh) \
+XX(tcpi_snd_cwnd, TcpiSndCwnd) \
+XX(tcpi_advmss, TcpiAdvmss) \
+XX(tcpi_reordering, TcpiReordering) \
+XX(tcpi_rcv_rtt, TcpiRcvRtt) \
+XX(tcpi_rcv_space, TcpiRcvSpace) \
+XX(tcpi_total_retrans, TcpiTotalRetrans) \
+XX(tcpi_pacing_rate, TcpiPacingRate) \
+XX(tcpi_max_pacing_rate, TcpiMaxPacingRate) \
+XX(tcpi_bytes_acked, TcpiBytesAcked) \
+XX(tcpi_bytes_received, TcpiBytesReceived) \
+XX(tcpi_segs_out, TcpiSegsOut) \
+XX(tcpi_segs_in, TcpiSegsIn) \
+XX(tcpi_notsent_bytes, TcpiNotsentBytes) \
+XX(tcpi_min_rtt, TcpiMinRtt) \
+XX(tcpi_data_segs_in, TcpiDataSegsIn) \
+XX(tcpi_data_segs_out, TcpiDataSegsOut) \
+XX(tcpi_delivery_rate, TcpiDeliveryRate) \
+XX(tcpi_busy_time, TcpiBusyTime) \
+XX(tcpi_rwnd_limited, TcpiRwndLimited) \
+XX(tcpi_sndbuf_limited, TcpiSndbufLimited) \
+XX(tcpi_delivered, TcpiDelivered) \
+XX(tcpi_delivered_ce, TcpiDeliveredCe) \
+XX(tcpi_bytes_sent, TcpiBytesSent) \
+XX(tcpi_bytes_retrans, TcpiBytesRetrans) \
+XX(tcpi_dsack_dups, TcpiDsackDups) \
+XX(tcpi_reord_seen, TcpiReordSeen)
```

Comment:
Please indent each `XX` two spaces. (Hint: you may want to use `clang-format`)

Reply:
Also, I guess this may backfire if the kernel does not support all these fields. Do we care?

Reply:
This is the client, so I guess we do since we can't control it.
According to the protocol specification, only a small subset of `tcp_info`'s fields are required. Perhaps I should just read those?